<div class="alert alert-success">
<h2> Комментарий ревьюера 4</h2>

Привет. Теперь обозначим, что это четвертая итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Привет. Оформление комментариев по работе сохраняется. Только обозначим, что это третья итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Привет еще раз. Спасибо, что доделала работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Привет, Наталья! Спасибо, что прислала задание:) Поздравляю с первым сданным проектом. Меня зовут Слепцов Артем и я буду проверять твой проект) Ты проделала большую работу над проектом, однако его еще можно улучшить. Будет здорово, если ты, надеюсь, не против, если я буду на ты, будешь отвечать на комментарии и участвовать в диалоге. 

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках (если фон комментария <font color=green>зелёный</font> - всё сделано правильно. Рекомендации укажу таким же цветом; <font color=orange>оранжевый</font> - некритичные замечания. Если таких замечаний меньше трех - проект может быть принят без их отработки; <font color=red>красный</font> - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Путь к файлу: /datasets/data.csv

## Описание данных:

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

Для того чтобы ответить на вопрос заказчика будет првоедена работа по следующим направления:
- подключить необходимые библитеки дял работы
- закрузить файл для работы с указанного пути хранения
- провести преобработку данных (удаление пропущенных значений, проверить типы данных, проверка и удаление дубликатов, состаление лемма, составить категоризаию данных)
- ответить на инетерсующиеся вопросы заказчика

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<h2> Комментарий ревьюера</h2>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе следует приводить краткий план того, что надо сделать(если этот план имеется), а также информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут. Также название работы лучше отображать в отдельной ячейке и делать крупный шрифт (заголовок). Так работа выглядит презентабельно. Подробнее смотри [методичку](https://www.notion.so/35d9c2f32ffa42f3b17010737aedca4b) по оформлению проектов.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Вступление в работу очень важно. Так ты поясняешь то, чему она посвящена. Цель работы также хорошо бы вынести в отдельный блок вступления. Еще лучше будет, если ты сделаешь план работы с гиперссылками. Как это реализовать - смотри по [ссылке](https://stackoverflow.com/questions/49535664/how-to-hyperlink-in-a-jupyter-notebook/49717704).

</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
from pymystem3 import Mystem
import collections

## Вывод

Подключение библиотек для работы с файлом. Далее необходимо поключить файл для работы. 

Путь к файлу '/datasets/data.csv'


In [2]:
data_credit = pd.read_csv('/datasets/data.csv')

print(data_credit.info())

print('================')
print('Вывод данных')
display(data_credit.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
Вывод данных


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-warning">
<h2> Комментарий ревьюера 2</h2>

Таблицы на печать выводи с помощью метода display. print переводит данные в строку, а нам этого не надо. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

👍  

</div>

## Вывод

12 столбцов и 21 525 записей. 

Нужно обратить внимание что количество значений в столбцах различается:
- days_employed       19351 non-null float64
- total_income        19351 non-null float64

Значит есть пропущенные, не заполненные записи. Необходимо провести предобработку данных.

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

У меня имеется несколько комментариев по данному шагу:
- желательно выводить около 10 строчек таблицы. Меньше не рекомендуется, можно не увидеть структуру данных;
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен.

</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
print('Кол-во пропусков')
print(data_credit.isnull().sum())

Кол-во пропусков
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


## Вывод

Выяснили кол-во пустых записей, проблема подтвердилась. 
В столбцах:
'days_employed'  и 'total_income' имеются не заполненные данные. 

Заполенние пропущенных данных будет проводится по следующему алгоритму:
- составим список с категориями "Тип занятости" (income_type) и "Возраст клиента в годах"(dob_years).
- вычислим медиану ежемесячного дохода(total_income) исходя из типа занятости (income_type), полученый показатель запишем в пропуски в столбце total_income.
- выичслим средний арифметический показтель стажа (days_employed ) исходя и возраста клиента (dob_years), полученный результат запишем в пропуски в столбе dob_years.

In [4]:
#Получим список имеющихся типы занятости
income_type = data_credit['income_type'].unique()

for i in income_type: #По каждому типу занятости
    #вычислим медиану ежемесячный доход. 
    med_total = data_credit.loc[data_credit['income_type'] == i, 'total_income'].median()
    
    #Полученную сумму в медианах пропишем в пропусках в столбце 'total_income'
    data_credit.loc[data_credit['income_type'] == i, 'total_income'] = data_credit.loc[data_credit['income_type'] == i, 'total_income'].fillna(med_total)

## Вывод

Вычислили медиану по всем типам занятости, провели запись в пропущенные записи.

В таблице ещё есть пустоты и их надо тоже вычислить и заполнить.

In [5]:
#Получим список имеющихся возраст клиентов в годах
dob_years = data_credit['dob_years'].unique()

for y in dob_years: # По каждому возрасту
    # вычислим среднеарифметический показатель стажа 
    mean_dob_years = data_credit.loc[data_credit['dob_years'] == y, 'days_employed'].mean()
    
    #Полученные среднеарифметические показатели стажа пропишем в пропусках в столбце 'days_employed'
    data_credit.loc[data_credit['dob_years'] == y, 'days_employed'] = data_credit.loc[data_credit['dob_years'] == y, 'days_employed'].fillna(mean_dob_years)

## Вывод

Вычислили среднеарифметический показатель стажа по всем возрастам клиета банка, провели запись в пропущенные записи.

Проверим результат и наличие пропущенных значений.

In [6]:
print('Кол-во пропусков')
print(data_credit.isnull().sum())

Кол-во пропусков
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


## Вывод

Обработка пропусков прошла успешно.


<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Обработать пропуски таким образом мы не можем. Данные у нас о клиентах, получивших крелиты. Лучше использовать заполнение пропусков на основе группировки данных по одному из параметров в нашей таблице. Подумай, какой параметр будет больше всего влиять на доход. От чего больше всего зависит доход? Проведи группировку по нему и на его основании заполни пропуски. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Пропуски заполнены. Конечно, тип занятости сильно влияет на получаемый доход. Здорово, что ты смогла реализовать такую обработку пропусков.  

</div>

### Замена типа данных

In [7]:
# список столбцов с указанием тип данных
print(data_credit.dtypes) 
# заменяем тип float на int
data_credit['days_employed']= data_credit['days_employed'].astype('int')
data_credit['total_income']= data_credit['total_income'].astype('int')

print('===================')
print('Результат после замены типа:')
print(data_credit.dtypes)

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object
Результат после замены типа:
children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int64
purpose             object
dtype: object


Вывод

Тип полей "days_employed" и "total_income" перевела в тип int, теперь будет точность в расчетах до целого числа.
Использоала метод .astype() для того чтобы отображалось целое число.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена верно. Также хорошо бы освоить метод to_numeric() для будущих работ. 

</div>

### Обработка дубликатов

In [8]:
#Все значения строковых типов перевела в нижний регистр
data_credit['education']=data_credit['education'].str.lower() 
data_credit['family_status']=data_credit['family_status'].str.lower() 
data_credit['purpose']=data_credit['purpose'].str.lower() 
data_credit['gender']=data_credit['gender'].str.lower() 
data_credit['income_type']=data_credit['income_type'].str.lower() 

#Суммарного количества дубликатов и удаление дубликатов
print('Число дубликатов равна', data_credit.duplicated().sum()) 
data_credit = data_credit.drop_duplicates().reset_index(drop=True)

print('===================')
print('Проверка на отсутствие дубликатов после обработки')
print('Число дубликатов равна', data_credit.duplicated().sum())

Число дубликатов равна 71
Проверка на отсутствие дубликатов после обработки
Число дубликатов равна 0


### Вывод

Все значения строковых типов перевела в нижний регистр, для того чтобы вычислить точное кол-во дубликатов в таблице. Суммарного количества дубликатов в таблице data_credit равна 71 запись. 
Удаление дубликатов проводила методом drop_duplicates().reset_index(drop=True), т.е после удаления индексы таблицы переназначились занова
После того как провели удаление дубликатов из таблицы, проверка показало 0 записей

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.
    
Вывод информации на печать лучше оформлять. Т.е. не выводить просто 71 и 0, а писать "Число дубликатов до/после удаления".

</div>

### Лемматизация

In [9]:
m = Mystem() 
#Создаю список категорий причины для лематизации
kategoriya = data_credit['purpose'].unique()
print(kategoriya)

def field_purpose(pur):
    kat = m.lemmatize(pur)
    if 'жилье' in kat:
        return 'недвижимость'
    if 'недвижимость' in kat:
        return 'недвижимость'
    if 'автомобиль' in kat:
        return 'автомобиль'
    if 'образование' in kat:
        return 'образование'
    if 'свадьба' in kat:
        return 'свадьба'  
    
#Создаем столбец "Категория кредита" и записываем результат функции
data_credit['kategoriya'] = data_credit['purpose'].apply(field_purpose)
display(data_credit.head(10))

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   среднее             1   
8         2          -6929         35    высшее             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      f   сотрудник     0        253875   
1   женат / замужем                 0      f   сотрудник     0        112080   
2   женат / замужем                 0      m   сотрудник     0        145885   
3   женат / замужем                 0      m   сотрудник     0        267628   
4  гражданский брак                 1      f   пенсионер     0        158616   
5  гражданский брак                 1      m   компаньон     0        255763   
6   женат / замужем                 0      f   компаньон     0        240525   
7   женат / замужем                 0      m   сотрудник     0        135823   
8  гражданский брак                 1      f   сотрудник     0         95856   
9   женат / замужем                 0      m   сотрудник     0        144425   

                      purpose    kategoriya  
0               покупка жилья  недвижимость  
1     приобретение автомобиля    автомобиль  
2               покупка жилья  недвижимость  
3  дополнительное образование   образование  
4             сыграть свадьбу       свадьба  
5               покупка жилья  недвижимость  
6           операции с жильем  недвижимость  
7                 образование   образование  
8       на проведение свадьбы       свадьба  
9     покупка жилья для семьи  недвижимость

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Таблицы на печать выводи с помощью метода display. print переводит данные в строку, а нам этого не надо. 

</div>

### Вывод

В поле "purpose" нету единного формата отображения причины, для быстрого анализа повода взятия кредита необходимо произвести  лемматизацию. Для лемматизации было использована библиотека Mystem с функцией pymystem3. 

Перед тем как провести лемму, в отдельную Series-таблицу Kategoriya с помощью выполения метода unique() я записала список цели получения кредита  из столбца "purpose"  из таблицы data_credit. 

После вывода на экран таблицы Kategoriya, я провела ручной анализ и выделила основные часто встречающиеся слова существительного части речи, необходимые для составления лемма. 
Это:
- "недвижимость", 
- "жилье", 
- "автомобиль", 
- "образование", 
- "свадьба"

Далее создаю функцию для проверки слов в сведённых к лемме.  Создаем дополнительный столбец "kategoriya" где записываются результаты проверок.


<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Круто, ты сделала самую сложную часть проекта. Использовала нужные методы, и даже провела группировку с помощью собственной функции. Только выбор лемм стоит объяснить подробнее.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Да, вывод уни кальных целей вполне достаточен для выбора главных целей кредитов. 

</div>

### Категоризация данных

In [10]:
# выведем типы education и их кол-во из таблицы, запишем в словарь
education = collections.Counter()
education_list = data_credit['education']
for i in education_list:
    education[i] += 1
print('Словарь "Уровень образования клиента"')   
print(education)


Словарь "Уровень образования клиента"
Counter({'среднее': 15172, 'высшее': 5250, 'неоконченное высшее': 744, 'начальное': 282, 'ученая степень': 6})


In [11]:
# выведем типы family_status и их кол-во из таблицы, запишем в словарь
family_status = collections.Counter()
family_status_list = data_credit['family_status']
for i in family_status_list:
    family_status[i] += 1
print('Словарь "Семейное положение"')      
print(family_status)

Словарь "Семейное положение"
Counter({'женат / замужем': 12339, 'гражданский брак': 4151, 'не женат / не замужем': 2810, 'в разводе': 1195, 'вдовец / вдова': 959})


In [12]:
# выведем типы income_type и их кол-во из таблицы, запишем в словарь
income_type = collections.Counter()
income_type_list = data_credit['income_type']
for i in income_type_list:
    income_type[i] += 1
print('Словарь "Тип занятости"')     
print(income_type)

Словарь "Тип занятости"
Counter({'сотрудник': 11084, 'компаньон': 5078, 'пенсионер': 3829, 'госслужащий': 1457, 'безработный': 2, 'предприниматель': 2, 'студент': 1, 'в декрете': 1})


### Вывод

Библиотека collections предназначена для удобного и быстрого подсчета количества появлений неизменяемых элементов в таблице. Результат посчета выводится словарем.

Я записала в словари по трём столбцам:
- education
- family_status
- income_type

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

В данных есть словари. Попробуй их обнаружить и выделить. Это довольно распространенный способ хранения информации. Мне важно понять, что он тобой освоен. Заметь, что в основной таблице мы имеем по 2 параметра с образованием и семейным статусом. Именно по ним и стоит выделить словари. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Словари имелись в виду несколько иные, а именно family_status_id:family_status. Так мы сможем убрать лишние текстовые параметры из основной таблицы, что увеличит эффективность работы с данными. Однако важно, что ты освоила данный способ хранения информации и умеешь его использовать. 

</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [13]:
#вывод группы по кол-ву детей у кредитора и подсчет долей по признаку debt
print(data_credit.pivot_table(index='children', columns='debt', values='purpose', aggfunc='count')*100/data_credit['children'].count())

debt              0         1
children                     
-1         0.214412  0.004661
 0        60.725273  4.954787
 1        20.341195  2.069544
 2         8.660390  0.904260
 3         1.412324  0.125851
 4         0.172462  0.018645
 5         0.041950       NaN
 20        0.316957  0.037289


### Вывод

Проверка показала наличие ошибочных и пропущенных данных в столбце "children". 
Проведу замену -1 на 1, 20 на 2, NaN на 0

In [14]:
#замена ошибочных значений -1 и 20
data_credit['children'] = data_credit['children'].replace(-1, 1)
data_credit['children'] = data_credit['children'].replace(20, 2)
#замена пропущенных значений NaN
data_credit['children'] = data_credit['children'].fillna(0)  

In [15]:
group = data_credit.pivot_table(index='children', columns='debt', values='purpose', aggfunc='count')
print(group)

# вычисляем долю зависимости наличие детей на возврат кредита в срок
print('Не имеют задолженность по возврату кредитов (в долях)')
print(group[0]/(group[1]+group[0]))
print('')
print('Имеют задолженность по возврату кредитов (в долях)')
print(group[1]/(group[1]+group[0]))


debt            0       1
children                 
0         13028.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN
Не имеют задолженность по возврату кредитов (в долях)
children
0    0.924562
1    0.908342
2    0.905075
3    0.918182
4    0.902439
5         NaN
dtype: float64

Имеют задолженность по возврату кредитов (в долях)
children
0    0.075438
1    0.091658
2    0.094925
3    0.081818
4    0.097561
5         NaN
dtype: float64


<div class="alert alert-danger">
<h2> Комментарий ревьюера 2</h2>

Категоризовать данные по числу детей необязательно. Разбиение есть уже в исходных данных. Все-таки разница в возврате кредита у семей с одним ребенком или тремя детьми будет. Так ты только уменьшаешь объем имеющейся информации. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Значения в -1 и 20 детей стоит обработать отдельно. Данные показатели выглядят малореальными. По поводу таблиц смотри комментарий ниже. 

</div>

<div class="alert alert-danger">
<h2> Комментарий на комментарий ревьюера 2</h2>

я уже незнаю что от меня требуется. кажется это последняя идея которая ко мне пришла.

</div>

### Вывод

Для того чтобы ответить на вопрос: Есть ли зависимость между наличием детей и возвратом кредита в срок?
Я разделила кредиторов на группы по кол-ву детей с отметкой по столбцу debt и разделила на кол-во кредитовов всего в группе.
## замену пропущенных значение NaN на 0 не сработало. где я не правильно сделала?

Результат показал наличие четверых детей (доля группы равна 0.097561) больше влияет на оплату кредита в срок, чем меньше детей тем легче возращать долги.

<div class="alert alert-success">
<h2> Комментарий ревьюера 4</h2>

Молодец, теперь ты рассчитываешь правильные доли. Смотри, это пропуски в сводной таблице, а не в исходной data_credit. Клиентов с долгом и с 5 детьми просто нет в данных, поэтому там и образуется пропуск. 

</div>

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Смотри, ты судишь по абсолютным значениям. Что это значит? Там, где больше людей имеют долг, те группы и опаснее для банка. Однако может быть так, что число должников в определенной группе 1000, а кредит брали 100000. Тогда процент должников в группе будет 1%. А в другой группе должников может быть 500, но кредит брали 1000 человек. Доля должников - 50%. Как думаешь, что невыгоднее для банка? Судить по абсолютным значениям мы не можем. Надо найти долю должников в каждой из групп, потом уже сравнить полученны результаты и сделать вывод о том, какая группа хуже всего возвращает кредит. 

</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [16]:
# Группировка столбцов 'family_status' и 'debt', количественный посчет 
print(data_credit.groupby(['family_status', 'debt'])['debt'].count())


family_status          debt
в разводе              0        1110
                       1          85
вдовец / вдова         0         896
                       1          63
гражданский брак       0        3763
                       1         388
женат / замужем        0       11408
                       1         931
не женат / не замужем  0        2536
                       1         274
Name: debt, dtype: int64


### Вывод

Для того чтобы ответить на вопрос: Есть ли зависимость между семейным положением и возвратом кредита в срок?
Я провела группировку столбцов 'family_status' и 'debt' и вывела количественный посчет

Наибольший результат показал, что зависимость между женат / замужем и возвратом кредита в срок есть.

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Так не пойдет. Не стоит рассчитывать показатели руками. Групп может быть и 100, так уже не рассчитать. Используй методы группировки groupby или pivot_table. 

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [17]:
#составляет квантилии данных по ежемесячным доходам
data_credit['qcut_total_income'] = pd.qcut(data_credit['total_income'],5)
# смотрим кол-во людей в каждой группе
a= data_credit['qcut_total_income'].value_counts()
print(a)
# смотрим кол-во по квантилиям с условием имел ли задолженность по возврату кредитов
print(data_credit.groupby(['qcut_total_income', 'debt'])['debt'].count())

(214618.2, 2265604.0]    4291
(161335.0, 214618.2]     4291
(98537.6, 132134.4]      4291
(20666.999, 98537.6]     4291
(132134.4, 161335.0]     4290
Name: qcut_total_income, dtype: int64
qcut_total_income      debt
(20666.999, 98537.6]   0       3947
                       1        344
(98537.6, 132134.4]    0       3930
                       1        361
(132134.4, 161335.0]   0       3915
                       1        375
(161335.0, 214618.2]   0       3930
                       1        361
(214618.2, 2265604.0]  0       3991
                       1        300
Name: debt, dtype: int64


<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Здорово, что ты освоила метод qcut. Так выделяемые группы получаются равного размера, и мы можем по ним построить выводы. 

</div>

### Вывод

Для того чтобы ответить на вопрос: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Я составила квантилии данных по ежемесячным доходам с помощью метода qcut(), даннные диапозона записала в новый столбец "qcut_total_income". 

Далее убелились что в каждой группе одинаковое кол-во людей (записей). Группировкой по новому столбцу "qcut_total_income" и отметкой в столбце "debt" получили количественный состав данных.

Т.к в самом маленьком диапозоне (20666.999, 98537.6] с отметкой имеет задолженность равна не самому минимальному кол-ву, то могу сказать что зависимости между уровнем дохода и возвратам кредита в срок нету.

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Исходя из чего выбран именно такой способ категоризации столбца с доходами? 2 группы - слишком мало для построения по ним надежных выводов. Лучше использовать разбиение по квантилям данных (метод qcut). Тогда группы получатся одинакового размера. Или же можно узнать в интернете процентный состав общества по доходам и затем уже категоризовать столбец с доходами по найденному соотношению с помощью персентилей. 

</div>

- Как разные цели кредита влияют на его возврат в срок?

In [18]:
kat=data_credit.pivot_table(index=['kategoriya'], columns='debt', values='purpose', aggfunc='count')
print(kat)
print('Не имеют задолженность по возврату кредитов')
print(kat[0]/(kat[1]+kat[0]))

print('')
print('Имеют задолженность по возврату кредитов')
print(kat[1]/(kat[1]+kat[0]))

debt              0    1
kategoriya              
автомобиль     3903  403
недвижимость  10029  782
образование    3643  370
свадьба        2138  186
Не имеют задолженность по возврату кредитов
kategoriya
автомобиль      0.906410
недвижимость    0.927666
образование     0.907800
свадьба         0.919966
dtype: float64

Имеют задолженность по возврату кредитов
kategoriya
автомобиль      0.093590
недвижимость    0.072334
образование     0.092200
свадьба         0.080034
dtype: float64


<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

На данном шаге рекомендуется использовать метод сводных таблиц (pivot_table()). Он является аналогом метода groupby(). Изучи метод сводных таблиц и примени его хотя бы в одном пункте анализа.

</div>

### Вывод

Для того чтобы ответить на вопрос: Как разные цели кредита влияют на его возврат в срок?
Я создала сводную таблицу методом pivot_table(), после чего наглядно видно кол-во целей кредита и погашен/не погашен в срок. Для расчета доли по категриям я взяла показатель категории разделила на сумму значений 0 и 1.

Наибольший результат показал, что цель кредита "автомобиль" больше всего влияет на возврат кредита в срок, доля составляет более 0.093590.


<div class="alert alert-danger">
<h2> Комментарий ревьюера 2</h2>

По каждому пункту анализа важно делать выводы не по абсолютным показателям, а по долям должников. По полученным сводным таблицам найди доли должников в группах. Выводы делай именно по долям должников. Сравнивать стоит именно доли. 

</div>

<div class="alert alert-danger">
<h2> Комментарий ревьюера 3</h2>

Ты находишь долю должников и долю недолжников относительно общего числа клиентов в данных. Эти показатели нормированные, но выводы ты все равно делаешь по числу должников. В категории недвижимость больше всего должников по твоей таблице, однако это не так. В данной категории будет маименьшая доля должников. Давай теперь уточню то, как ее рассчитывать. Смотри, таблица (модернизация того, что у тебя выше) - `data_credit.pivot_table(index=['kategoriya'], columns='debt', values='purpose', aggfunc='count')` дает нам число клиентов по категориям и дает показатели по числу должников и недолжников в них. Для того чтобы найти доли должников в группах нам надо поделить их число на общее число клиентов (должники + недолжники). Для этого нам надо взять столбец 0 и поделить его на сумму столбцов 0 и 1. Затем уже мы можем сделать выводы по полученным долям. Это и будут искомые доли по числу должников в группах.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 4</h2>

Доли рассчитаны правильно. Здорово, что ты разобралась. Только по анализу влияния семейного статуса на возврат кредита также стоило рассчитать доли должников. 

</div>

### Шаг 4. Общий вывод

Полученный от заказчика файл сотсоит из:
- 12 столбцов 
- 21 525 записей с данными.

В ручном просмотре файла было замечано:
- Пропущенные данные. Файл по этой проблеме прошел обработку
- Отрицательные данные. Но они не влияли на то, чтобы ответить на главные вопросы заказчика. Поэтому обработку мною проведена не была
- Замена тип данных. float на integer
- Удаление дубликатов. 
- Лемматизация категорий причины взятия кредита. ДОбавили дополнительный столбце с категорией кредита.
- Категоризация данных. Составление словарей и подсчет кол-ва ключей в таблице


Итоговой работой над файлом было то, чтобы ответить на важные вопросы:

1. Есть ли зависимость между наличием детей и возвратом кредита в срок?

Для того чтобы ответить на вопрос я разделила кредиторов на группы по кол-ву детей с отметкой по столбцу debt и разделила на кол-во кредитовов всего в группе.

Результат показал наличие четверых детей (доля группы равна 0.097561) больше влияет на оплату кредита в срок, чем меньше детей тем легче возращать долги. 


Статистика результата:

- Не имеют задолженность по возврату кредитов (в долях)

Кол-во детей/ доля показателя
- 0    0.924562
- 1    0.908342
- 2    0.905075
- 3    0.918182
- 4    0.902439
- 5        1


- Имеют задолженность по возврату кредитов (в долях)

Кол-во детей/ доля показателя
- 0    0.075438
- 1    0.091658
- 2    0.094925
- 3    0.081818
- 4    0.097561
- 5         0

Вывод наличие детей на возврат кредит в срок не значителен.


2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

Провела группировку категории семейного положения и количественный подсчет по признаку "Имеет ли задолжность по возрату кредита".

Результат проверки показал, что:
- в разводе, не имеет задолжности по кредиту кол-во кредитов составляет 1110
- в разводе, имеет задолжности по кредиту кол-во кредитов составляет 85
- вдовец / вдова, не имеет задолжности по кредиту кол-во кредитов составляет 896
- вдовец / вдова, имеет задолжности по кредиту кол-во кредитов составляет 63
- гражданский брак, не имеет задолжности по кредиту кол-во кредитов составляет 3763
- гражданский брак, имеет задолжности по кредиту кол-во кредитов составляет 388
- женат / замужем, не имеет задолжности по кредиту кол-во кредитов составляет 11408
- женат / замужем имеет задолжности по кредиту кол-во кредитов составляет 931
- не женат / не замужем, не имеет задолжности по кредиту кол-во кредитов составляет 2536
- не женат / не замужем, имеет задолжности по кредиту кол-во кредитов составляет 274

Нас интересуют данные, которые находятся с отметкой "имеет ли задолжность по возрату кредита".  Большее кол-во покозало при семейном положении "женат / замужем " - 931


3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Я составила квантилии данных по ежемесячным доходам, даннные диапозона записала в новый столбец "qcut_total_income". Получилось 5 групп входящие в диапозон с доходами. Группировкой по новому столбцу "qcut_total_income" и отметкой в столбце "debt" получили количественный состав данных.

Минимальный диапозон с доходами (20666.999, 98537.6] и с отметкой имеет задолжность по возврату кредита (1) составляет 344 записей.

Когда как диапозон с доходами (132134.4, 161335.0]  и с отметкой имеет задолжность по возврату кредита (1) составляет 375 записей.

Поэтому, можно судить что наименьший уровень дохода не влияет на возврат кредита в срок.


4. Как разные цели кредита влияют на его возврат в срок?

Составила сводную таблицу по столбцу "Категория кредита" и по признаку "Имеет ли задолжность по возрату кредита".

Для расчета доли по категориям я взяла показатель категории разделила на сумму значений 0 и 1.

Наибольший результат показал, что цель кредита "автомобиль" больше всего влияет на возврат кредита в срок, доля составляет более 0.093590.

Результат проверки показал, следующее:

kategoriya/ не имеет задолжность:                      
                  
- автомобиль   0.906410
- недвижимость  0.927666
- образование   0.907800
- свадьба       0.919966 
 

 kategoriya/ имеет задолжность:                      
                  
- автомобиль    0.093590
- недвижимость  0.072334
- образование   0.092200
- свадьба       0.080034

Вывод: Автомобиль имеет больший покатель- 0.093590 в задолжности по оплате кредита.

Итак, мы в итоге получается:
1. Наибольший результат показал, что зависимость между детьми и возвратом кредита в срок есть.
2. Наибольший результат показал, что зависимость между женат / замужем и возвратом кредита в срок есть.
3. Наименьший размер ежемесячного дохода не влияет на возврат кредита в срок.
4. Наибольший результат показал, что цель кредита "автомобиль" влияет на возврат кредита в срок есть.



<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. В нем можно приводить полученные в ходе работы значения. Также можно расписать все, что было сделано в работе.

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, правильно. Здорово, что по каждому пункту работы приведен вывод.     

</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

- В начале работы следует описывать данные, с которыми работаешь;
- Попробуй улучшить обработку пропусков;
- Категоризацию по доходу стоит делать иначе. Смотри данный пункт подробнее;
- Выдели словари в данных;
- Используй методы группировки при подсчетах и ответах на поставленные в задании вопросы (как минимум в одном пункте примени метод сводных таблиц);
- Финальный вывод стоит писать подробно.

Ты проделала большую работу, молодец! Исправь отмеченные выше помарки. Также есть и желтые комментарии. Думаю, исправить их тебе не составит труда. Жду твою работу :)

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Ты очень здорово поработала, молодец. Осталось лишь исправить последний пункт при ответе на поставленные в задании вопросы. Как сделаешь это, присылай проект снова :) 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Я верю в то, что ты справишься с расчетами. Остался буквально последний шаг. Рассчитай нужные доли, исправь выводы там, где это необходимо и присылай проект. Буду ждать :)

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 4</h2>

Помарки исправлены, и теперь работа выполнена хорошо. Я рад, что ты справилась с задачей. Ты проделала большую работу! Успехов в дальнейших проектах :)

</div>
   

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

In [19]:
print('Проектная часть выполнена. С новым годом!')

Проектная часть выполнена. С новым годом!


<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Тебя также с наступающим новым годом! Успехов во всех начинаниях.

</div>